api/v1beta1/gcpmachine_types.go

```diff
@@ -146,6 +149,65 @@ const (
 	HostMaintenancePolicyTerminate HostMaintenancePolicy = "Terminate"
 )
 
+// KeyType is a type for disk encryption.
+type KeyType string
+
+const (
+	// CustomerManagedKey (CMEK) references an encryption key stored in Google Cloud KMS.
+	CustomerManagedKey KeyType = "Managed"
+	// CustomerSuppliedKey (CSEK) specifies an encryption key to use.
+	CustomerSuppliedKey KeyType = "Supplied"
+)
+
+// ManagedKey is a reference to a key managed by the Cloud Key Management Service.
+type ManagedKey struct {
+	// KMSKeyName is the name of the encryption key that is stored in Google Cloud KMS. For example:
+	// "kmsKeyName": "projects/kms_project_id/locations/region/keyRings/key_region/cryptoKeys/key
+	// +kubebuilder:validation:Required
+	KMSKeyName string `json:"kmsKeyName,omitempty"`
```

Comment:
Does GCP specify any constraints for the name of the KMS key? Upper case? Lower case? Certain special characters allowed? MInimum or maximum length? All of this could be validated at admission time to prevent errors down the line

Reply:
So unless I'm missing something I don't see constraints defined in the API https://pkg.go.dev/google.golang.org/api/compute/v1#CustomerEncryptionKey or in validations. I'm leery to add additional limit checks here unless its well defined.

Reply:
When trying to create a key, it gives me
```
Key names can contain letters, numbers, underscores (_), and hyphens (-). Keys can't be renamed or deleted.
```
I believe that is also true of project IDs and regions so we probably can limit to that selection plus the slashes required

Reply:
Thanks for testing that @JoelSpeed.

Reply:
Thanks, I'll add a check in the webhook.Validator.

Reply:
Any webhook validations also should be explained in the godoc

Personal quarrel but, webhook validations are hard to review from an API review perspective. We have kubebuilder validations and CEL validations that can handle these validations and live alongside the field, it's much easier to see the validation and understand the behaviour of the field when it's written right next to it.

I know the CAPI projects lean a lot on webhooks and they seem preferred for now, but, I want to raise that point at a future community call

Reply:
OK, so would be preferred way for something like this to use:
// +kubebuilder:validation:Pattern
and
// +kubebuilder:validation:MaxLength
?

Reply:
I think that's clearer personally yes.

You can also do CEL stuff but I know CAPI aren't doing that much yet, for this one a pattern probably makes sense as it's simple to understand

```
// +kubebuilder:validation:pattern="projects/[-_[a-z0-9]+/locations/region/keyRings/[-_[a-z0-9]+/cryptoKeys/[-_[a-z0-9]+"
```

For the other field you could do:

```
// +kubebuilder:validation:XValidation:rule="self.contains('@')",message="kmsKeyServiceAccount should be an email address"
// +kubebuilder:validation:XValidation:rule="self.split('@')[0].matches('[a-z][-a-zA-Z0-9]{5,29}')",message="kmsServiceAccount email prefix must start with a lower case letter, followed by one or more lower case alphanumerical characters that can be separated by hyphens and must be between 6 and 30 characters in length.
// +kubebuilder:validation:XValidation:rule="self.endsWith('.iam.gserviceaccount.com')",message="kmsKeyServiceAccount must be a valid gcloud service account. Must end with '.iam.gserviceaccount.com'."
```

Will want a maintainers ack on using these though, I'm not a maintainer here

Reply:
^ cc. @richardcase @cpanato

Reply:
Thanks for the guidance. With some minor changes that first regexp seems to work well, e.g. https://regex101.com/r/7Zjc3O/1

For the kmsKeyServiceAccount, I had to reduce the number of Validation rules because of test failures
`Forbidden: estimated rule cost exceeds budget by factor of more than 100x (try simplifying the rule, or adding maxItems, maxProperties, and maxLength) `. 
I derived the maxlength from the google docs, e.g. https://cloud.google.com/iam/quotas

Reply:
@JoelSpeed - you make a good point about kubebuilder validations/CEL vs webhook validation. I think part of the current reliance on webhook validation in CAPI is historical because CEL wasn't available for more complex validation logic.

Reply:
Sounds good. Hopefully we can get this merged and revisit the validations after the discussion.